src/server/pkg/storage/gc/util.go

```diff
@@ -0,0 +1,152 @@
+	db.LogMode(false)
+	db.DB().SetMaxOpenConns(3)
+	db.DB().SetMaxIdleConns(2)
+	if err := initializeDb(db); err != nil {
+		return nil, err
+	}
+	return db, nil
+}
+
+// WithLocalDB creates a local database client for testing during the lifetime of
+// the callback.
+func WithLocalDB(f func(*gorm.DB) error) (retErr error) {
+	db, err := NewLocalDB()
+	if err != nil {
+		return err
+	}
+	if err := clearData(db); err != nil {
+		return err
+	}
+	defer func() {
+		if err := clearData(db); retErr == nil {
+			retErr = err
+		}
+	}()
+	return f(db)
+}
+
+func clearData(db *gorm.DB) error {
+	if err := db.Exec("DELETE FROM chunks *").Error; err != nil {
+		return err
+	}
+	return db.Exec("DELETE FROM refs *").Error
+
+}
+
+// WithLocalGarbageCollector creates a local garbage collector client for testing during the lifetime of
+// the callback.
+func WithLocalGarbageCollector(f func(context.Context, obj.Client, Client) error, opts ...Option) error {
+	return obj.WithLocalClient(func(objClient obj.Client) error {
+		return WithLocalDB(func(db *gorm.DB) error {
+			return WithGarbageCollector(objClient, db, func(ctx context.Context, client Client) error {
+				return f(ctx, objClient, client)
+			}, opts...)
+		})
+	})
+}
+
+// WithGarbageCollector creates a garbage collector client for testing during the lifetime of
+// the callback.
+func WithGarbageCollector(objClient obj.Client, db *gorm.DB, f func(context.Context, Client) error, opts ...Option) error {
+	client, err := NewClient(db)
+	if err != nil {
+		return err
+	}
+	// (bryce) may want to pipe a real context through here.
+	cancelCtx, cancel := context.WithCancel(context.Background())
+	eg, gcContext := errgroup.WithContext(cancelCtx)
+	eg.Go(func() error {
+		return Run(gcContext, objClient, db, opts...)
+	})
+	eg.Go(func() error {
+		defer cancel()
+		return f(gcContext, client)
+	})
+	return eg.Wait()
+}
+
+func isRetriableError(err error) bool {
+	if err, ok := err.(*pq.Error); ok {
+		return err.Code.Class().Name() == "transaction_rollback"
+	}
+	return false
+}
+
+type statementFunc func(*gorm.DB) *gorm.DB
+
+func runTransaction(ctx context.Context, db *gorm.DB, name string, stmtFuncs []statementFunc) error {
+	for {
+		err := collectSQLStats(name, func() error {
+			return tryTransaction(ctx, db, stmtFuncs)
+		})
+		if isRetriableError(err) {
+			continue
+		}
+		return err
+	}
+}
+
+func tryTransaction(ctx context.Context, db *gorm.DB, stmtFuncs []statementFunc) error {
+	txn := db.BeginTx(ctx, &sql.TxOptions{Isolation: sql.LevelSerializable})
+
+	// So we don't leak in case of a panic somewhere unexpected
+	defer func() {
+		if r := recover(); r != nil {
+			txn.Rollback()
+		}
+	}()
+
+	if err := txn.Error; err != nil {
+		return err
+	}
+
+	for _, stmtFunc := range stmtFuncs {
+		if err := stmtFunc(txn).Error; err != nil {
+			txn.Rollback()
+			return err
+		}
+	}
+
+	if err := txn.Commit().Error; err != nil {
+		txn.Rollback()
+		return err
+	}
+
+	return nil
```

Comment:
Minor nit, but it's a little weird that this function has so much more vertical whitespace than the other ones. I'd remove the empty lines here.

Reply:
Sure, this was one of the few untouched places from Marc's work.